### PR TITLE
Reject Batches with Invalid Characters

### PR DESCRIPTION
**Pull request by ives1227**
**Rejects the batch and sends an email**
* * *

**Ticket Link**: https://jira.huit.harvard.edu/browse/LTSEPADD-117

# How should this be tested?
1. Add the two new variables to your already existing .env (or set up a new .env from env-example.txt):
```
RELATIVE_DRS_CONFIG_PATH=data/blobs/sidecarfiles/drsConfig.txt
INVALID_CHARACTERS_MESSAGE="Invalid characters were found in the export.  Please see documentation (forthcoming) on what to do when this occurs."
```

2. Run the tests as described here: https://github.com/harvard-lts/epadd-curator-app#running-tests

# Test coverage
Yes/No: Are changes in this pull-request covered by:
- unit tests? yes
- integration tests? no

### CODE REVIEW

File: app/monitor_exports/__init__.py [hunk 1]
```diff
@@ -348,16 +351,43 @@ def collect_exports():
...
+                            #Get the failure email
+                            message = traceback.format_exc()
+                            logger.error(message)
+                            email = send_invalid_character_email(manifest_parent_prefix)
+                            break
                     break
 
     return manifest_object_keys
 
+def prepare_and_send_invalid_character_email(manifest_parent_prefix):
```
Comment (awoods) [inline app/monitor_exports/__init__.py:372]: Is this method used anywhere?

File: scripts/monitor_epadd_exports.py [hunk 2]
```diff
@@ -36,10 +36,12 @@ def main():
     try: 
         # Collect exports
         export_object_keys = monitor_exports.collect_exports()
-    except:
-        logging.error(traceback.format_exc())
+    except Exception as e:
+        message = traceback.format_exc()
+        logging.error(message)
```
Comment (awoods) [inline scripts/monitor_epadd_exports.py:41]: For future reference, here is another way to have the logger print the local exception:
https://github.com/harvard-lts/jstor_aspace_publisher/blob/main/app/jstor_publisher.py#L152-L153

See: https://docs.python.org/3/library/logging.html#logging.Logger.debug

**Review by awoods — CHANGES_REQUESTED** [review]
Looks good.

I get the following error when running the tests within the container:
```
======================================================================
ERROR: test_invalid_character_email (scripts.unit_tests.TestSendNotification)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/appuser/epadd-curator-app/scripts/unit_tests.py", line 229, in test_invalid_character_email
    message = monitor_epadd_exports.send_invalid_character_email(local_directory)
  File "/home/appuser/epadd-curator-app/app/monitor_exports/__init__.py", line 382, in send_invalid_character_email
    payload_data = construct_payload_body(download_export_path, manifest_parent_prefix)
  File "/home/appuser/epadd-curator-app/app/monitor_exports/__init__.py", line 206, in construct_payload_body
    raise MonitoringException("drsConfig.txt not found in export {}".format(download_dir), None)
monitor_exports.monitor_exception.MonitoringException: drsConfig.txt not found in export /home/appuser/epadd-curator-app/download_exports/

----------------------------------------------------------------------
Ran 13 tests in 24.890s

FAILED (errors=1)
```

File: app/monitor_exports/__init__.py (hunk 1 above)
Comment (ives1227) [reply app/monitor_exports/__init__.py:372]: Yes - fixed - it was two methods but for testing I had to break them in half.

File: scripts/monitor_epadd_exports.py (hunk 2 above)
Comment (ives1227) [reply scripts/monitor_epadd_exports.py:41]: Thanks

**Comment by ives1227** [comment]
> Looks good.
> 
> I get the following error when running the tests within the container:
> 
> ```
> ======================================================================
> ERROR: test_invalid_character_email (scripts.unit_tests.TestSendNotification)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/appuser/epadd-curator-app/scripts/unit_tests.py", line 229, in test_invalid_character_email
>     message = monitor_epadd_exports.send_invalid_character_email(local_directory)
>   File "/home/appuser/epadd-curator-app/app/monitor_exports/__init__.py", line 382, in send_invalid_character_email
>     payload_data = construct_payload_body(download_export_path, manifest_parent_prefix)
>   File "/home/appuser/epadd-curator-app/app/monitor_exports/__init__.py", line 206, in construct_payload_body
>     raise MonitoringException("drsConfig.txt not found in export {}".format(download_dir), None)
> monitor_exports.monitor_exception.MonitoringException: drsConfig.txt not found in export /home/appuser/epadd-curator-app/download_exports/
> 
> ----------------------------------------------------------------------
> Ran 13 tests in 24.890s
> 
> FAILED (errors=1)
> ```

@awoods this is fixed. A file did not get committed because of the gitignore

**Review by awoods — APPROVED** [review]
Changes and tests look great.

**Pull request merged**